### PR TITLE
MINIFICPP-1357 Implement and test proxy handling in InvokeHTTP

### PR DESCRIPTION
**Pull request by lordgamez**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/http-curl/processors/InvokeHTTP.cpp [hunk 1]
```diff
@@ -246,6 +246,20 @@ void InvokeHTTP::onSchedule(const std::shared_ptr<core::ProcessContext> &context
   if (context->getProperty(DisablePeerVerification.getName(), disablePeerVerification)) {
     utils::StringUtils::StringToBool(disablePeerVerification, disable_peer_verification_);
   }
+
+  std::string proxy_value;
+  if (context->getProperty(ProxyHost.getName(), proxy_value) && !proxy_value.empty()) {
+    proxy_.host = proxy_value;
+  }
+  if (context->getProperty(ProxyPort.getName(), proxy_value) && !proxy_value.empty()) {
+    proxy_.port = std::stoi(proxy_value);
```
Comment (adamdebreceni) [inline extensions/http-curl/processors/InvokeHTTP.cpp:255]: `getProperty` is a templated method, that depending on the second argument's type carries out checks and conversions from `std::string`, so feel free to pass `proxy.port` as the second argument here, moreover I believe `getProperty` adheres to the contract "the out argument is not modified on failure, missing value", so we don't even need the "if"s here

File: libminifi/include/utils/HTTPClient.h [hunk 2]
```diff
@@ -300,6 +319,9 @@ class BaseHTTPClient {
   virtual void setUploadCallback(HTTPUploadCallback *callbackObj) {
   }
 
+  virtual void setSeekFunction(HTTPUploadCallback *callbackObj) {
+  }
```
Comment (adamdebreceni) [inline libminifi/include/utils/HTTPClient.h:323]: it seems like `HTTPClient` is the only class inheriting from it, and noone directly instantiates a BaseHTTPClient, could we make these noop virtual methods pure virtual?

File: extensions/http-curl/processors/InvokeHTTP.cpp (hunk 1 above)
Comment (arpadboda) [reply extensions/http-curl/processors/InvokeHTTP.cpp:255]: The other reason to remove if statements is updating the config: we should null the members, otherwise proxy properties can't be removed without restarting minifi.

**Comment by hunyadi-dev** [comment]
Please check the CI errors, they seem related to the change.

File: libminifi/include/utils/ByteArrayCallback.h [hunk 3]
```diff
@@ -57,6 +57,7 @@ class ByteInputCallBack : public InputStreamCallback {
   }
 
   virtual void seek(size_t pos) {
+    ptr = &vec[pos];
```
Comment (szaszm) [inline libminifi/include/utils/ByteArrayCallback.h:60]: What's the purpose of this change? I think this may break `getBuffer`.

Comment (lordgamez) [reply libminifi/include/utils/ByteArrayCallback.h:60]: Curl proxy authentication requires 2 steps (getting the available authentication methods then authenticate with the chosen one) and after the first step it needs a data rewind which requires implementing the seek function. You are right, it broke the getBuffer that's what partly caused the test failures, it is fixed in [fd3a7c6](https://github.com/apache/nifi-minifi-cpp/pull/897/commits/fd3a7c653202c7cdd460153175d3deb13b7651ce)

File: extensions/http-curl/processors/InvokeHTTP.h [hunk 4]
```diff
@@ -155,8 +155,10 @@ class InvokeHTTP : public core::Processor {
   bool use_chunked_encoding_{false};
   // penalize on no retry
   bool penalize_no_retry_{false};
-  // disable peer verification ( makes susceptible for MITM attacks )
-  bool disable_peer_verification_{false};
+  bool disable_peer_verification_{false}; // ( makes susceptible for MITM attacks )
+  bool follow_redirects_{true};
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply extensions/http-curl/processors/InvokeHTTP.h]: Fixed in  [e0e31c5](https://github.com/apache/nifi-minifi-cpp/pull/897/commits/e0e31c5e96e0ad29c54ca09230a845588664b864)

File: docker/test/integration/test_http.py [hunk 5]
```diff
@@ -34,3 +33,26 @@ def test_invoke_listen():
...
+
+    listen_flow = ListenHTTP(8080) >> LogAttribute() >> PutFile('/tmp/output')
+
+    with DockerTestCluster(SingleFileOutputValidator('test')) as cluster:
+        cluster.put_test_data('test')
+        cluster.deploy_flow(None, engine='http-proxy')
+        cluster.deploy_flow(listen_flow, name='minifi-listen')
+        cluster.deploy_flow(invoke_flow, name='minifi-invoke')
+
+        assert cluster.check_output()
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/test_http.py:58]: We could check the acces logs of the http-proxy, that's what I was checking manually, I will add it to the test.

File: libminifi/include/utils/HTTPClient.h (hunk 2 above)
Comment (lordgamez) [reply libminifi/include/utils/HTTPClient.h]: Done in [fd3a7c6](https://github.com/apache/nifi-minifi-cpp/pull/897/commits/fd3a7c653202c7cdd460153175d3deb13b7651ce)

File: extensions/http-curl/processors/InvokeHTTP.cpp (hunk 1 above)
Comment (lordgamez) [reply extensions/http-curl/processors/InvokeHTTP.cpp]: Done in [2bc2b75](https://github.com/apache/nifi-minifi-cpp/pull/897/commits/429675f5d9a526c529080f4468d170e320f21dca)

File: libminifi/include/utils/ByteArrayCallback.h [hunk 6]
```diff
@@ -65,7 +66,7 @@ class ByteInputCallBack : public InputStreamCallback {
   }
 
   virtual char *getBuffer(size_t pos) {
-    return ptr + pos;
+    return reinterpret_cast<char*>(&vec[0]) + pos;
```
Comment (szaszm) [inline libminifi/include/utils/ByteArrayCallback.h:69]: Please use `vec.data()` to avoid out-of-bounds indexing when `vec` is empty, or index with pos.
```suggestion
    return &vec[pos];
```
or
```suggestion
    return vec.data() + pos;
```

Comment (lordgamez) [reply libminifi/include/utils/ByteArrayCallback.h]: Done in [839ed47](https://github.com/apache/nifi-minifi-cpp/pull/897/commits/839ed47d2c3618a1fb1443f0929bfc8908051a87)

File: docker/test/integration/test_http.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/test_http.py:58]: Added check in [5047001](https://github.com/apache/nifi-minifi-cpp/pull/897/commits/504700148228b571ee69a2d58322c235d5f71428)

**Review by szaszm — APPROVED** [review]

**Review by adamdebreceni — APPROVED** [review]
